src/app/execution/txRunner.js

```diff
@@ -43,150 +27,179 @@ TxRunner.prototype.rawRun = function (args, cb) {
-          number: self.blockNumber,
-          coinbase: coinbases[self.blockNumber % coinbases.length],
-          difficulty: difficulties[self.blockNumber % difficulties.length],
-          gasLimit: new BN(gasLimit, 10).imuln(2)
-        },
-        transactions: [],
-        uncleHeaders: []
-      })
-      if (!args.useCall) {
-        ++self.blockNumber
-      } else {
-        executionContext.vm().stateManager.checkpoint()
+    var warnEstimation = " An important gas estimation might also be the sign of a problem in the contract code. Please check loops and be sure you did not sent value to a non payable function (that's also the reason of strong gas estimation)."
+    if (gasEstimation > gasLimit) {
+      return callback('Gas required exceeds limit: ' + gasLimit + '. ' + warnEstimation)
+    }
+    if (gasEstimation > blockGasLimit) {
+      return callback('Gas required exceeds block gas limit: ' + gasLimit + '. ' + warnEstimation)
+    }
+
+    tx.gas = gasEstimation
+
+    if (self._api.config.getUnpersistedProperty('doNotShowTransactionConfirmationAgain')) {
+      return executeTx(tx, null, self._api, callback)
+    }
+    self._api.detectNetwork((err, network) => {
+      if (err) {
+        console.log(err)
+        return
+      }
+      if (network.name !== 'Main') {
+        return executeTx(tx, null, self._api, callback)
       }
 
-      executionContext.vm().runTx({block: block, tx: tx, skipBalance: true, skipNonce: true}, function (err, result) {
-        if (args.useCall) {
-          executionContext.vm().stateManager.revert(function () {})
+      var amount = executionContext.web3().fromWei(typeConversion.toInt(tx.value), 'ether')
+      var content = confirmDialog(tx, amount, gasEstimation, self,
+        (gasPrice, cb) => {
+          let txFeeText, priceStatus
+          // TODO: this try catch feels like an anti pattern, can/should be
+          // removed, but for now keeping the original logic
+          try {
+            var fee = executionContext.web3().toBigNumber(tx.gas).mul(executionContext.web3().toBigNumber(executionContext.web3().toWei(gasPrice.toString(10), 'gwei')))
+            txFeeText = ' ' + executionContext.web3().fromWei(fee.toString(10), 'ether') + ' Ether'
+            priceStatus = true
+          } catch (e) {
+            txFeeText = ' Please fix this issue before sending any transaction. ' + e.message
+            priceStatus = false
+          }
+          cb(txFeeText, priceStatus)
+        },
+        (cb) => {
+          executionContext.web3().eth.getGasPrice((error, gasPrice) => {
+            var warnMessage = ' Please fix this issue before sending any transaction. '
+            if (error) {
+              return cb('Unable to retrieve the current network gas price.' + warnMessage + error)
+            }
+            try {
+              var gasPriceValue = executionContext.web3().fromWei(gasPrice.toString(10), 'gwei')
+              cb(null, gasPriceValue)
```

Comment:
seems the awaited callback is more something like `cb(txFeeText, gasPriceValue, gasPriceStatus)`

Reply:
correct, so this call is valid, I prefer to avoid passing another null as a 3rd parameter if it's not necessary (much like for a typicall callback cb(err, result) when there is an error only cb(err) is necessary instead of cb(err, null) )